Automattic/jetpack
# Hijack the feature eligibility checks on WordPress.com sites

Fixes https://github.com/Automattic/wp-calypso/issues/51593.
Fixes https://github.com/Automattic/wp-calypso/issues/51826.
Fixes https://github.com/Automattic/wp-calypso/issues/61499.

#### Changes proposed in this Pull Request:
Hijacks the `Jetpack_Plan::supports` function with a call to `wpcom_site_has_feature` (if exists) to make sure that Jetpack features are gated correctly on WP.com sites.

That allows us to simplify some of the existing feature eligibility checks since we can use now `Jetpack_Plan::supports` on all sites.

#### Jetpack product discussion
pdG1ka-46-p2

#### Does this pull request change what data or activity we track or use?
N/A.

#### Testing instructions:

Jetpack sites:
- [Spin up a JN site using this branch](https://jurassic.ninja/create?jetpack-beta&branches.jetpack=update/feature-checks-wpcom-sites).
- Set up Jetpack and keep it on the Free plan.
- Apply D75906-code to your WP.com sandbox.
- Enable incoming HTTP connections in your WP.com sandbox: `sudo global-http enable`.
- Go to Settings > Jetpack Constants and enter your WP.com sandbox address in `JETPACK__SANDBOX_DOMAIN`.
- Go to Jetpack > Settings > Performance and enable VideoPress.
- Go to Posts > Add new.
- Insert a Video block and upload a video.
- Make sure it's correctly processed by VideoPress. 
- Insert a Payment Button block.
- Make sure it requires a plan upgrade.

Simple sites:
- Apply D75906-code to your WP.com sandbox.
- Sandbox the API and a WP.com simple site with a Free plan.
- Go to https://wordpress.com and switch to your sandboxed site.
- Go to Posts > Add new.
- Make sure these blocks require a Personal plan:
  - Payments button
  - Donations
  - Premium Content
- Make sure these blocks require a Premium plan: 
  - Pay with PayPal
  - Video
  - Calendly
  - OpenTable
  - WhatsApp button
- Upgrade to a Business plan.
- Make sure the blocks above no longer require any upgrade.
- Insert a Tiled Gallery block in a post and publish it.
- Visit the published post.
- Make sure the Tiled Gallery block is rendered as expected.

Atomic sites:
- Create a WoA dev blog with a Business plan (p9o2xV-1r2-p2).
- Install Jetpack Beta and activate the branch of this PR.
- Go to Jetpack > Search.
- Make sure the instant search experience requires an upgrade.
- Click on the upsell nudge to buy the upgrade.
- Make sure the instant search experience is available now.
- Go to Posts > Add new.
- Insert a Video block and upload a video.
- Make sure it's correctly processed by VideoPress. 
- Insert a Payment Button block.
- Make sure it does not require any upgrade.
- Add this change to wpcomsh (you can use SFTP via Settings > Hosting Config) to simulate that the site no longer has a Business plan (*):
```diff
diff --git a/class.atomic-plan-manager.php b/class.atomic-plan-manager.php
index d8abaf8..b179fdd 100644
--- a/class.atomic-plan-manager.php
+++ b/class.atomic-plan-manager.php
@@ -73,6 +73,7 @@ class Atomic_Plan_Manager {
 	public static function current_plan_slug() {
+		return self::FREE_PLAN_SLUG;
 		$atomic_site_id = wpcomsh_get_atomic_site_id();
 
diff --git a/wpcom-features/functions-wpcom-features.php b/wpcom-features/functions-wpcom-features.php
index c542764..c15ce1d 100644
--- a/wpcom-features/functions-wpcom-features.php
+++ b/wpcom-features/functions-wpcom-features.php
@@ -72,6 +72,8 @@ function wpcom_site_has_feature( $feature, $blog_id = 0 ) {
 		$is_wpcom_site = is_blog_wpcom( $blog ) || is_blog_atomic( $blog );
 	}
 
+	$purchases = array();
+
 	return WPCOM_Features::has_feature( $feature, $purchases, $is_wpcom_site );
 }
 
```

- Go to `wp-admin/post-new.php`.
- Make sure the Video block requires a Premium plan and that the upload buttons are disabled.
- Make sure the Payment Button block requires a Personal plan.
- Go to `wp-admin/upload.php`.
- Make sure you cannot upload video files.

(*) I first tried this by actually removing the plan from SA, but that left the site on a broken state because Jetpack was deactivated. Didn't look deeply into the issue, but apparently it's caused by a conflict between Jetpack Beta and WoA dev blogs since I can remove a plan without problems when Jetpack Beta is not involved.
